### DISCUSSION TITLE
Supervision Edges Attributes in LinkNeighborLoader

### DISCUSSION THREAD

**Discussion by YoavLotem** [pyg-team/pytorch_geometric#5133] [Ideas]
Hey!

I'm working on an edge classification problem.
I'm using the `LinkNeighborLoader `to sample supervision edges and their corresponding labels.

Since the supervision edges (defined by `edge_label_index`) are not necessarily part of the message passing edges (defined by `edge_index`) I'm having a hard time adding their features.

I tried concatenating them to the `edge_label` and extract them in the training loop but I got a "ValueError: only one element tensors can be converted to Python scalars", which I guess means the code assumes the `edge_label` has 1-d.

Maybe it could be beneficial to add a `label_edge_attr` parameter to the `LinkNeighborLoader`.

Can think of any workaround to solve this problem?

**Comment by rusty1s** [comment]
Interesting. So it looks like you want to integrate edge-level features that are only available for supervision edges. Is that correct? What is the use-case of this? I think for now concatenating is a valid approach. Can you share an reproducible example of the error? This does not seem intended.

**Reply by YoavLotem** [reply]
Hi, thanks for your quick response.

So in terms of relevance - this could useful in every situation in which the supervision edges are not part of the message passing edges - therefore their features could not be part of `edge_attr`.

For example, if we want to model users of a website as nodes of a graph when edges are different types of interactions (transactions, social network friends etc.) and predict a property ("bad-transaction", "worked together", etc.) of that connection, we might add the other interaction features (transaction amount, number of shared Facebook friends) to the final prediction model in the training loop.

**Reply by YoavLotem** [reply]
Reproducible Example:

import torch
from torch_geometric.data import Data
from torch_geometric.loader import LinkNeighborLoader
import torch_geometric.transforms as T


edge_index = torch.tensor([[0, 1, 1, 2, 0, 1, 2],
                           [1, 0, 2, 1, 3, 3, 3]], dtype=torch.long)

x = torch.tensor([[-1], [0], [1], [4], [3]], dtype=torch.float)

edge_attr = torch.tensor([[1.0], [2.0], [1.0], [1.0], [1.0], [1.0], [1.0]], dtype=torch.float)

edge_label_index = torch.tensor([[0, 1, 2],
                                 [4, 4, 4]], dtype=torch.long)


edge_attr_supervision = torch.tensor([[1.0], [2.0], [3.0]], dtype=torch.float)

data = Data(x=x, edge_index=edge_index, edge_attr=edge_attr)


### No Problem
edge_lbl = torch.tensor([0, 1, 0], dtype=torch.long)  
loader = LinkNeighborLoader(data=data, num_neighbors=[2, 2], edge_label_index=edge_label_index, edge_label=edge_lbl)
print(next(iter(loader)))

#### Data(x=[4, 1], edge_index=[2, 3], edge_attr=[3, 1], edge_label_index=[2, 1], edge_label=[1])

### Cause Error
edge_lbl = torch.cat([edge_lbl.unsqueeze(1), edge_attr_supervision], dim=1)
loader = LinkNeighborLoader(data=data, num_neighbors=[2, 2], edge_label_index=edge_label_index, edge_label=edge_lbl)
print(next(iter(loader)))

#### ValueError: only one element tensors can be converted to Python scalars

**Reply by YoavLotem** [reply]
@rusty1s kind reminder

**Reply by rusty1s** [reply]
Yes, will look into this ASAP, sorry for the delay!

**Reply by rusty1s** [reply]
I just pushed a fix via https://github.com/pyg-team/pytorch_geometric/pull/5186. Thanks for being patient!

**Reply by YoavLotem** [reply]
Thank you very much! I appreciate it.

I hope your team would consider my initial suggestion - I think that many real world applications might benefit from it.

**Reply by rusty1s** [reply]
Definitely, sorry for the "hacky" workaround for now :(
